### PR TITLE
Workaround for APFS DMG resizing

### LINKED ISSUE #117: hdiutil: resize: failed. Invalid argument (22)
My machine is now running macOS Ventura, but I could also reproduce the issue on Monterey.

Steps to reproduce:
1. Clone the repo
2. Run `./sshrd.sh 16.1`

The script will fail with an error `hdiutil: resize: failed. Invalid argument (22)`

This appears to be caused by this line in `sshrd.sh`:

`hdiutil resize -size 210MB work/ramdisk.dmg`

The README says that we can't resize the APFS DMGs on Linux machines, but it should work on macOS?

### PR DESCRIPTION
**Pull request by chrisspankroy**
Fixes #117 

Since we can't resize the APFS DMG, we can create a bigger blank DMG, use dd to copy everything from the old one, then expand the APFS container in the new DMG.

This seems to work, but I am unable to boot this project on my iOS 16.1 iPad to verify. I run `./sshrd.sh boot` and the screen lights up for a second but then turns off again and SSHRD never boots.

Does this project work on iOS 16.1 or does it need other fixes?

### CODE REVIEW

**Comment by chrisspankroy** [comment]
I can confirm that this fixes the issue with APFS ramdisks, I can successfully boot a ramdisk created this way

**Comment by ranzhie07** [comment]
Error while apfs resizing container
image

**Comment by ranzhie07** [comment]
> Fixes #117 
> 
> 
> 
> Since we can't resize the APFS DMG, we can create a bigger blank DMG, use dd to copy everything from the old one, then expand the APFS container in the new DMG.
> 
> 
> 
> This seems to work, but I am unable to boot this project on my iOS 16.1 iPad to verify. I run `./sshrd.sh boot` and the screen lights up for a second but then turns off again and SSHRD never boots.
> 
> 
> 
> Does this project work on iOS 16.1 or does it need other fixes?

Im having error on this
diskutil apfs resizeContainer /dev/disk3 210m

Error
/dev/disk3 is neither an APFS Container Reference nor an APFS Physical Store

Can u help me?

**Comment by ranzhie07** [comment]
> Is the ramdisk you're trying to expand using APFS?
> 
> If not, I suppose we could check if the DMG is APFS rather than check if the iOS version is >= 16.1. That's probably the better check to do anyway in case it changes in the future 

Its apfs bro

**Comment by chrisspankroy** [comment]
Comment out line 235 (the eject line) and try running again and then post the output of sshrd.sh and `sudo diskutil list`

**Comment by ranzhie07** [comment]
> Comment out line 235 (the eject line) and try running again and then post the output of the script again and the output of `sudo diskutil list`

I have comment 235 line same error
While running command
diskutil apfs resizeContainer $ourmountpoint 210m 

Error: 
image

Heres the output of diskutil list
image

**Comment by chrisspankroy** [comment]
Maybe `resizeContainer` should be called on the actual container instead of the disk itself. That doesn't explain why this works as it is on my end though...maybe Ventura's `diskutil` is smarter than Big Sur's `diskutil` and can sort itself out.

Assuming /dev/disk3 and /dev/disk4 are still as they are in your output of `diskutil list` above, try running `diskutil apfs resizeContainer /dev/disk4 210m` and tell me what the output is

**Comment by ranzhie07** [comment]
> Maybe `resizeContainer` should be called on the actual container instead of the disk itself. That doesn't explain why this works as it is on my end though...maybe Ventura's `diskutil` is smarter than Big Sur's `diskutil` and can sort itself out.
> 
> 
> 
> Assuming /dev/disk3 and /dev/disk4 are still as they are in your output of `diskutil list` above, try running `diskutil apfs resizeContainer /dev/disk4 210m` and tell me what the output is

Still no luck bro here's the error output

image

**Comment by chrisspankroy** [comment]
Interesting. And I'm guessing a `diskutil apfs resizeContainer /dev/disk3 210m` yields the same error as the script?

**Comment by ranzhie07** [comment]
> Interesting. And I'm guessing a `diskutil apfs resizeContainer /dev/disk3 210m` yields the same error as the script?

No bro id i do on disk3
Error is
/dev/disk3 is neither an APFS Container reference nor an APFS Physical Store.
image

When i try on disk4
Error is 

image

**Comment by ranzhie07** [comment]
> Fixes #117 
> 
> 
> 
> Since we can't resize the APFS DMG, we can create a bigger blank DMG, use dd to copy everything from the old one, then expand the APFS container in the new DMG.
> 
> 
> 
> This seems to work, but I am unable to boot this project on my iOS 16.1 iPad to verify. I run `./sshrd.sh boot` and the screen lights up for a second but then turns off again and SSHRD never boots.
> 
> 
> 
> Does this project work on iOS 16.1 or does it need other fixes?

Do you figureout the problem bro?

File: sshrd.sh [hunk 1]
```diff
@@ -212,7 +212,35 @@ else
...
+            diskutil unmountDisk $ourmountpoint
+            dd if=work/ramdisk.dmg of=$ourmountpoint bs=1m
+
+            diskutil apfs resizeContainer $ourmountpoint 210m
+
+            # Exit if resize fails
+            [ $? -ne 0 ] && echo "Failed to resize container" && exit
+
+            # Now we have a bigger ramdisk, eject and replace
+            diskutil eject $ourmountpoint
```
Comment (ICH525) [inline sshrd.sh:235]: sudo diskutil list

**Comment by ICH525** [comment]
still no luck guys with making ramdisk

**Comment by ICH525** [comment]
krnl
dtre
rtsc
rdsk
iOS 16+ requested
created: /Users/hasnit3ch/Desktop/SSHRD_Script/work/temp.dmg
Mountpoint detected as /dev/disk6
Unmount of all volumes on disk6 was successful
98+0 records in
98+0 records out
102760448 bytes transferred in 1.863045 secs (55157256 bytes/sec)
Password:
/dev/disk6 is neither an APFS Container Reference nor an APFS Physical Store
[-] An error occurred



NO luck same with Catalina

**Comment by chrisspankroy** [comment]
Can you try it on Ventura? I can try testing on an older macOS version to see if I can reproduce

**Comment by ICH525** [comment]
The problem is i don't have a Ventura.. I did it with sudo apfs resizeContainer that doesn't do the work.. I tried to search and dig into it how te resize Container didn't work out.. It would be better when you work on older version like Catalina.

**Comment by ICH525** [comment]
For example After creating a hdiutil apfs that name as disk2 and can we create another apfs like disk2s1into that disk2? I think we can resize Container into that disk2s1... Confirm me if that can be work or not i really don't know how to create inside that disk2 as disk2s1...

**Comment by ICH525** [comment]
> Fixes #117
> 
> Since we can't resize the APFS DMG, we can create a bigger blank DMG, use dd to copy everything from the old one, then expand the APFS container in the new DMG.
> 
> This seems to work, but I am unable to boot this project on my iOS 16.1 iPad to verify. I run `./sshrd.sh boot` and the screen lights up for a second but then turns off again and SSHRD never boots.
> 
> Does this project work on iOS 16.1 or does it need other fixes?

the result of this is here

iOS 16+ requested
created: /Users/terminalcreeds/SSHRD_Script/work/temp.dmg
Mountpoint detected as /dev/disk2
Unmount of all volumes on disk2 was successful
98+0 records in
98+0 records out
102760448 bytes transferred in 3.653126 secs (28129456 bytes/sec)
/dev/disk2 is neither an APFS Container Reference nor an APFS Physical Store
[-] An error occurred


this is on montery

image

**Review by naiff654321 — APPROVED** [review]

**Review by Lin1111Oo — APPROVED** [review]

**Comment by naiff654321** [comment]
Darwin/irecovery

**Review by naiff654321 — APPROVED** [review]

**Review by naiff654321 — APPROVED** [review]

**Comment by naiff654321** [comment]
> Fixes #117 
> 
> 
> 
> Since we can't resize the APFS DMG, we can create a bigger blank DMG, use dd to copy everything from the old one, then expand the APFS container in the new DMG.
> 
> 
> 
> This seems to work, but I am unable to boot this project on my iOS 16.1 iPad to verify. I run `./sshrd.sh boot` and the screen lights up for a second but then turns off again and SSHRD never boots.
> 
> 
> 
> Does this project work on iOS 16.1 or does it need other fixes?

**Review by naiff654321 — APPROVED** [review]

**Review by naiff654321 — APPROVED** [review]
Darwin/irecovery

**Comment by verygenericname** [comment]
fixed in a slightly different manner

**Pull request closed without merging**